prow/plugins/dco/dco.go

```diff
@@ -0,0 +1,361 @@
+// 2. Check the existing status context value
+// 3. Check the existing PR labels
+// 4. If signed off, apply appropriate labels and status context.
+// 5. If not signed off, apply appropriate labels and status context and add a comment.
+func handle(gc gitHubClient, cp commentPruner, log *logrus.Entry, org, repo string, pr github.PullRequest, addComment bool) error {
+	l := log.WithField("pr", pr.Number)
+
+	commitsMissingDCO, err := checkCommitMessages(gc, l, org, repo, pr.Number)
+	if err != nil {
+		l.WithError(err).Infof("Error running DCO check against commits in PR")
+		return err
+	}
+
+	existingStatus, err := checkExistingStatus(gc, l, org, repo, pr.Head.SHA)
+	if err != nil {
+		l.WithError(err).Infof("Error checking existing PR status")
+		return err
+	}
+
+	hasYesLabel, hasNoLabel, err := checkExistingLabels(gc, l, org, repo, pr.Number)
+	if err != nil {
+		l.WithError(err).Infof("Error checking existing PR labels")
+		return err
+	}
+
+	return takeAction(gc, cp, l, org, repo, pr, commitsMissingDCO, existingStatus, hasYesLabel, hasNoLabel, addComment)
+}
+
+func markdownSHAList(org, repo string, list []github.GitCommit) string {
+	lines := make([]string, len(list))
+	lineFmt := "- [%s](https://github.com/%s/%s/commits/%s) %s"
+	for i, commit := range list {
+		if commit.SHA == nil {
+			continue
+		}
+		// if we somehow encounter a SHA that's less than 7 characters, we will
+		// just use it as is.
+		shortSHA := *commit.SHA
+		if len(shortSHA) > 7 {
+			shortSHA = shortSHA[:7]
+		}
+		message := ""
+		if commit.Message != nil {
+			message = *commit.Message
+		}
+
+		// get the first line of the commit
+		message = strings.Split(message, "\n")[0]
+
+		lines[i] = fmt.Sprintf(lineFmt, shortSHA, org, repo, *commit.SHA, message)
+	}
+	return strings.Join(lines, "\n")
+}
+
+// shouldPrune finds comments left by this plugin.
+func shouldPrune(log *logrus.Entry, botName string) func(github.IssueComment) bool {
+	return func(comment github.IssueComment) bool {
+		if comment.User.Login != botName {
+			return false
+		}
+		return strings.Contains(comment.Body, dcoMsgPruneMatch)
+	}
+}
+
+func handlePullRequestEvent(pc plugins.PluginClient, pe github.PullRequestEvent) error {
+	return handlePullRequest(pc.GitHubClient, pc.CommentPruner, pc.Logger, pe)
+}
+
+func handlePullRequest(gc gitHubClient, cp commentPruner, log *logrus.Entry, pe github.PullRequestEvent) error {
+	org := pe.Repo.Owner.Login
+	repo := pe.Repo.Name
+
+	// we only reprocess on label, unlabel, open, reopen and synchronize events
+	// this will reduce our API token usage and save processing of unrelated events
+	switch pe.Action {
+	case github.PullRequestActionLabeled,
+		github.PullRequestActionUnlabeled,
+		github.PullRequestActionOpened,
+		github.PullRequestActionReopened,
+		github.PullRequestActionSynchronize:
+	default:
+		return nil
+	}
+
+	shouldComment := pe.Action == github.PullRequestActionSynchronize || pe.Action == github.PullRequestActionOpened
```

Comment:
Typically we treat 'reopen' actions the same way we do 'open' actions to make it easy to retrigger bot automation on a PR in the rare case of a dropped webhook or something other transient error.